leaderelection/context.go

```diff
@@ -0,0 +1,225 @@
+
+		le, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
+			Lock:          rl,
+			LeaseDuration: b.lec.LeaseDuration,
+			RenewDeadline: b.lec.RenewDeadline,
+			RetryPeriod:   b.lec.RetryPeriod,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(context.Context) {
+					logger.Infof("%q has started leading %q", rl.Identity(), bkt.Name())
+					la.Promote(bkt, enq)
+				},
+				OnStoppedLeading: func() {
+					logger.Infof("%q has stopped leading %q", rl.Identity(), bkt.Name())
+					la.Demote(bkt)
+				},
+			},
+			ReleaseOnCancel: true,
+
+			// TODO: use health check watchdog, knative/pkg#1048
+			Name: b.lec.Component,
+		})
+		if err != nil {
+			return nil, err
+		}
+		// TODO: use health check watchdog, knative/pkg#1048
+		// if lec.WatchDog != nil {
+		// 	lec.WatchDog.SetLeaderElection(le)
+		// }
+		buckets = append(buckets, &runUntilCancelled{le: le})
+	}
+	return &runAll{les: buckets}, nil
+}
+
+// unopposedElector promotes when run without needing to be elected.
+type unopposedElector struct {
+	bkt reconciler.Bucket
+	la  reconciler.LeaderAware
+	enq func(reconciler.Bucket, types.NamespacedName)
+}
+
+// Run implements Elector
+func (ue *unopposedElector) Run(ctx context.Context) {
+	ue.la.Promote(ue.bkt, ue.enq)
+}
+
+type runAll struct {
+	les []Elector
+}
+
+// Run implements Elector
+func (ra *runAll) Run(ctx context.Context) {
+	sg := sync.WaitGroup{}
+	defer sg.Wait()
+
+	for _, le := range ra.les {
+		sg.Add(1)
+		go func(le Elector) {
+			defer sg.Done()
+			le.Run(ctx)
+		}(le)
+	}
+}
+
+// runUntilCancelled wraps a single-term Elector into one that runs until
+// the passed context is cancelled.
+type runUntilCancelled struct {
+	le Elector
+}
+
+// Run implements Elector
+func (ruc *runUntilCancelled) Run(ctx context.Context) {
+	// Run executes a single iteration of an election:
+	// 1. Wait to become leader
+	// 2. Call OnStartedLeading
+	// 3. Wait for ctx cancel or loss of leader.
+	// 4. Call OnStoppedLeading
+	//
+	// We effectively want this in a loop, where the termination
+	// condition is cancellation of our context.
+	for {
+		ruc.le.Run(ctx)
+		select {
+		case <-ctx.Done():
+			return // Run quit because context was cancelled, we are done!
+		default:
+			// Context wasn't cancelled, start over.
+		}
+	}
+}
+
+type bucket struct {
+	component string
+	name      string
+
+	// We are bucket {index} of {total}
+	index uint32
+	total uint32
+}
+
+var _ reconciler.Bucket = (*bucket)(nil)
+
+// Name implements reconciler.Bucket
+func (b *bucket) Name() string {
+	// The resource name is the lowercase:
+	//   {component}.{workqueue}.{index}-of-{total}
+	return strings.ToLower(fmt.Sprintf("%s.%s.%02d-of-%02d", b.component, b.name, b.index, b.total))
+}
+
+// Has implements reconciler.Bucket
+func (b *bucket) Has(nn types.NamespacedName) bool {
+	h := fnv.New32a()
+	h.Write([]byte(nn.Namespace + "." + nn.Name))
+	ii := h.Sum32() % b.total
```

Comment:
This is going to be bad when `total` is power of 2 — you basically discard upper `32-log(total)` bits.
We can probably specialize that case
```
if b.total & (b.total-1) == 0 {
  hash = hash ^ (hash >> bits(b.total))
}
```

Reply:
Alternatives are also if we want to do some kind of consistent hashing, so during resize the objects don't migrate much from bucket to bucket — especially important for autoscaler. Less for others.
If we're doing online resizing.

Reply:
As discussed I have a more complicated solution, which we'll put in once this merges.